backend/services/user_groups.go

```diff
@@ -0,0 +1,441 @@
+		err := db.Get(&userID, sq.Select("id").From("users").Where(sq.Eq{"slug": userSlug}))
+		if err != nil {
+			return backend.WrapError("Unable to get user id from slug", backend.BadInputErr(err, fmt.Sprintf(`No user with slug %s was found`, userSlug)))
+		}
+
+		var userGroupMap []models.UserGroupMap
+		err = db.Select(&userGroupMap, sq.Select("*").
+			From("group_user_map").
+			Where(sq.Eq{
+				"user_id":  userID,
+				"group_id": groupID,
+			}))
+		if err != nil {
+			return backend.WrapError("Unable to group from group id", backend.DatabaseErr(err))
+		}
+		if len(userGroupMap) == 0 {
+			_, err = db.Insert("group_user_map", map[string]interface{}{
+				"user_id":  userID,
+				"group_id": groupID,
+			})
+			if err != nil {
+				return backend.WrapError("Unable to connect user to group", backend.DatabaseErr(err))
+			}
+		}
+	}
+
+	return nil
+}
+
+func CreateUserGroup(ctx context.Context, db *database.Connection, i CreateUserGroupInput) (*dtos.UserGroup, error) {
+	if err := isAdmin(ctx); err != nil {
+		return nil, backend.WrapError("Unwilling to create a user group", backend.UnauthorizedReadErr(err))
+	}
+
+	cleanSlug := SanitizeSlug(i.Slug)
+	if cleanSlug == "" {
+		return nil, backend.BadInputErr(errors.New("Unable to create operation. Invalid operation slug"), "Slug must contain english letters or numbers")
+	}
+
+	err := db.WithTx(context.Background(), func(tx *database.Transactable) {
+		id, _ := tx.Insert("user_groups", map[string]interface{}{
+			"slug": cleanSlug,
+			"name": i.Name,
+		})
+		if len(i.UserSlugs) > 0 {
+			AddUsersToGroup(tx, i.UserSlugs, id)
+		}
+	})
+
+	if err != nil {
+		return nil, backend.WrapError("Error creating user group", backend.BadInputErr(err, "A user group with this slug already exists; please choose another name"))
+	}
+	return &dtos.UserGroup{
+		Slug: i.Slug,
+		Name: i.Name,
+	}, nil
+}
+
+func ModifyUserGroup(ctx context.Context, db *database.Connection, i ModifyUserGroupInput) (*dtos.UserGroup, error) {
+	if err := isAdmin(ctx); err != nil {
+		return nil, backend.WrapError("Unwilling to modify a user group", backend.UnauthorizedReadErr(err))
+	}
+
+	if err := i.validateUserGroupInput(); err != nil {
+		return nil, backend.WrapError("Unable to modify user group", backend.BadInputErr(err, "Unable to modify user group due to bad input"))
+	}
+
+	userGroup, err := lookupUserGroup(db, i.Slug)
+	if err != nil {
+		return nil, backend.WrapError("Unable to modify user group", backend.UnauthorizedWriteErr(err))
+	}
+
+	err = db.WithTx(context.Background(), func(tx *database.Transactable) {
+		if i.Name != "" {
+			tx.Update(sq.Update("user_groups").Set("name", i.Name).Where(sq.Eq{"id": userGroup.ID}))
+		}
+		if len(i.UsersToRemove) > 0 {
+			for _, userSlug := range i.UsersToRemove {
+				var userID int64
+				tx.Get(&userID, sq.Select("id").From("users").Where(sq.Eq{"slug": userSlug}))
+				tx.Delete(sq.Delete("group_user_map").Where(sq.Eq{"user_id": userID, "group_id": userGroup.ID}))
```

Comment:
I need to double check, but you might be able to combine these queries. The sql might look like:
```sql
DELETE FROM group_user_map JOIN users on group_user_map.user_id = users.id WHERE users.slug=?
```

Reply:
oh of course! Yeah I think that looks correct; I'll try it out

Reply:
changed